File: analysis/import/CSVImporter/src/test/java/de/maibornwolff/codecharta/importer/csv/CSVRowTest.java
package de.maibornwolff.codecharta.importer.csv;

import com.google.common.collect.ImmutableMap;
import org.junit.Test;

import java.util.Map;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.hasSize;

public class CSVRowTest {
    private static final char PATH_SEPARATOR = '\\';

    @Test
    public void getPath_should_return_path_from_specified_path_column() {
        // given
        String name = "someNodeName";
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "path", "attrib"});
        String[] csvRow = new String[]{"projectName", "blubb2", "blubb3", name};

        // when
        CSVRow row = new CSVRow(csvRow, header, PATH_SEPARATOR);

        // then
        assertThat(row.getPath(), is(name));
    }

    @Test
    public void getFileName_should_return_filename_from_specified_path_column() {
        // given
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "Path", "attrib"});
        Map<String,String> nameExpectedFilenameMap = ImmutableMap.of(
                "someNodeName", "someNodeName",
                "someDir\\someName", "someName",
                "someDir\\anotherDir\\anotherName", "anotherName");

        for (String name : nameExpectedFilenameMap.keySet()) {
            // when
            CSVRow row = new CSVRow(new String[]{"projectName", "blubb2", "blubb3", name}, header, PATH_SEPARATOR);
            // then
            assertThat(row.getFileName(), is(nameExpectedFilenameMap.get(name)));
        }
    }

    @Test
    public void getFolderWithFile_should_return_folderWithFile_from_specified_path_column() throws Exception {
        // given
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "PATH", "attrib"});
        Map<String,String> nameExpectedFolderWithFileMap = ImmutableMap.of(
                "someNodeName", "",
                "someDir\\someName", "someDir\\",
                "someDir\\anotherDir\\anotherName", "someDir\\anotherDir\\");

        for (String name : nameExpectedFolderWithFileMap.keySet()) {
            // when
            CSVRow row = new CSVRow(new String[]{"projectName", "blubb2", "blubb3", name}, header, PATH_SEPARATOR);
            // then
            assertThat(row.getFolderWithFile(), is(nameExpectedFolderWithFileMap.get(name)));
        }
    }

    @Test
    public void getAttributes_should_read_float_from_metric_column() throws Exception {
        // given
        CSVHeader header = new CSVHeader(new String[]{"path", "attrib"});
        String[] rawRow = new String[]{"name", "3,2"};

        // when
        CSVRow row = new CSVRow(rawRow, header, PATH_SEPARATOR);

        // then
        assertThat(row.getAttributes().keySet(), hasSize(1));
        assertThat(row.getAttributes().get("attrib"), is(3.2f));
    }

    @Test
    public void getAttributes_should_ignore_string_in_metric_column() throws Exception {
        // given
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "path", "attrib"});
        String[] rawRow = new String[]{"projectName", "blubb2", "blubb3", "name", "3bla"};

        // when
        CSVRow row = new CSVRow(rawRow, header, PATH_SEPARATOR);

        // then
        assertThat(row.getAttributes().keySet(), hasSize(0));
    }

    @Test
    public void getAttributes_should_ignore_column_if_no_attributeName_in_head() throws Exception {
        // given
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "path"});
        String[] rawRow = new String[]{"projectName", "blubb2", "blubb3", "name", "3,0"};

        // when
        CSVRow row = new CSVRow(rawRow, header, PATH_SEPARATOR);

        // then
        assertThat(row.getAttributes().keySet(), hasSize(0));
    }

    @Test
    public void getAttributes_should_ignore_column_if_not_present_in_row() throws Exception {
        // given
        CSVHeader header = new CSVHeader(new String[]{"path", "head2", "head3", "missingValueColumn"});
        String[] rawRow = new String[]{"somePath", "1,2", "1.3"};

        // when
        CSVRow row = new CSVRow(rawRow, header, PATH_SEPARATOR);

        // then
        assertThat(row.getAttributes().keySet(), hasSize(2));
    }

    @Test(expected = IllegalArgumentException.class)
    public void should_throw_exception_if_no_path_column_present() {
        // given
        CSVHeader header = new CSVHeader(new String[]{"head1", "head2", "head3", "path"});

        // when
        new CSVRow(new String[]{"", ""}, header, PATH_SEPARATOR);

        // then throw
    }
}

File: analysis/import/CSVImporter/src/test/kotlin/de/maibornwolff/codecharta/importer/csv/CSVRowTest.kt
package de.maibornwolff.codecharta.importer.csv

import de.maibornwolff.codecharta.model.Path
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.hasItem
import org.hamcrest.CoreMatchers.not
import org.hamcrest.CoreMatchers.nullValue
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.hasSize
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import kotlin.test.assertFailsWith

private const val PATH_SEPARATOR = '\\'

class CSVRowTest : Spek({

    describe("Using a header with path column") {
        val header = CSVHeader(arrayOf("head1", "head2", "head3", "path", "attrib", "attrib2", ""))

        it("name of node should be filename from this columnn") {
            val nameExpectedFilenameMap = mapOf(
                Pair("someNodeName", "someNodeName"),
                Pair("someDir\\someName", "someName"),
                Pair("someDir\\anotherDir\\anotherName", "anotherName")
            )

            for (name in nameExpectedFilenameMap.keys) {
                val node = CSVRow(arrayOf("projectName", "blubb2", "blubb3", name), header, PATH_SEPARATOR).asNode()
                assertThat(node.name, `is`(nameExpectedFilenameMap[name]))
            }
        }

        it("path in Tree should be absolute file name from this column") {
            val nameExpectedFolderWithFileMap = mapOf(
                Pair("someNodeName", Path.TRIVIAL),
                Pair("someDir\\someName", Path(listOf("someDir"))),
                Pair("someDir\\anotherDir\\anotherName", Path(listOf("someDir", "anotherDir")))
            )

            for (name in nameExpectedFolderWithFileMap.keys) {
                val path = CSVRow(arrayOf("projectName", "blubb2", "blubb3", name), header, PATH_SEPARATOR).pathInTree()
                assertThat(path, `is`<Path>(nameExpectedFolderWithFileMap[name]))
            }
        }

        it("should throw exception if no path column present") {
            assertFailsWith(IllegalArgumentException::class) {
                CSVRow(arrayOf("", ""), header, PATH_SEPARATOR)
            }
        }

        it("should ignore columns if no attribute name in header") {
            val rawRow = arrayOf<String?>("1", "2", "3", "file", "4", "5", "6", "7")
            val node = CSVRow(rawRow, header, PATH_SEPARATOR).asNode()
            assertThat(node.attributes.keys, hasSize(5))
        }

        it("should ignore column if not in row") {
            val rawRow = arrayOf<String?>("blubb1", "blubb2", "blubb3", "path")
            val node = CSVRow(rawRow, header, PATH_SEPARATOR).asNode()
            assertThat(node.attributes.keys, not(hasItem("attrib")))
        }

        it("should have attribute for metric columns") {
            val rawRow = arrayOf<String?>("3,2", "2", "3", "file")
            val node = CSVRow(rawRow, header, PATH_SEPARATOR).asNode()
            assertThat(node.attributes["head1"] as Double, `is`<Double>(3.2))
        }

        it("should have NO attribute for non-metric columns") {
            val rawRow = arrayOf<String?>("bla", "2", "3", "file")
            val node = CSVRow(rawRow, header, PATH_SEPARATOR).asNode()
            assertThat(node.attributes["head1"], nullValue())
        }
    }
})
